File: src/main/java/no/nav/pensjon/selvbetjeningopptjening/util/LocalDateTimeFromEpochDeserializer.java
package no.nav.pensjon.selvbetjeningopptjening.util;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.deser.std.StdDeserializer;

import java.io.IOException;
import java.time.Instant;
import java.time.LocalDate;

import static no.nav.pensjon.selvbetjeningopptjening.util.DateUtil.ZONE_ID;

public class LocalDateTimeFromEpochDeserializer extends StdDeserializer<LocalDate> {

    public LocalDateTimeFromEpochDeserializer() {
        super(LocalDate.class);
    }

    @Override
    public LocalDate deserialize(JsonParser parser, DeserializationContext context) throws IOException {
        var instant = Instant.ofEpochMilli(parser.readValueAs(Long.class));
        return instant.atZone(ZONE_ID).toLocalDate();
    }
}

File: src/main/java/no/nav/pensjon/selvbetjeningopptjening/tech/json/LocalDateTimeFromEpochDeserializer.kt
package no.nav.pensjon.selvbetjeningopptjening.tech.json

import no.nav.pensjon.selvbetjeningopptjening.util.DateUtil
import tools.jackson.core.JsonParser
import tools.jackson.databind.DeserializationContext
import tools.jackson.databind.deser.std.StdDeserializer
import java.time.Instant
import java.time.LocalDate

/**
 * Serializes epoch milliseconds as java.time.LocalDate.
 */
class LocalDateTimeFromEpochDeserializer() : StdDeserializer<LocalDate>(Long::class.java) {

    override fun deserialize(parser: JsonParser, provider: DeserializationContext?): LocalDate =
        Instant.ofEpochMilli(parser.readValueAs(Long::class.java))
            .atZone(DateUtil.ZONE_ID)
            .toLocalDate()
}